myia/parser.py

```diff
@@ -311,13 +381,73 @@ def process_Return(self, block: 'Block', node: ast.Return) -> 'Block':
-            block.write(node.targets[0].id, anf_node)
-        else:
-            raise NotImplementedError(node.targets)
+        targ, = node.targets
+
+        def write(targ, anf_node):
+
+            if isinstance(targ, ast.Name):
+                # CASE: x = value
+                anf_node.debug.name = targ.id
+                block.write(targ.id, anf_node)
+
+            elif isinstance(targ, ast.Tuple):
+                # CASE: x, y = value
+                for i, elt in enumerate(targ.elts):
+                    op = self.environment.map(operator.getitem)
+                    new_node = Apply([op, anf_node, Constant(i)], block.graph)
+                    write(elt, new_node)
+
+            elif isinstance(targ, ast.Subscript):
+                if isinstance(targ.value, ast.Name):
+                    # CASE: x[y] = value
+                    op = self.environment.map(operator.setitem)
+                    obj = self.process_node(block, targ.value)
+                    idx = self.process_node(block, targ.slice)
+                    new_node = Apply([op, obj, idx, anf_node], block.graph)
+                    write(targ.value, new_node)
+                else:
+                    # UNSUPPORTED: f()[x] = value
+                    raise NotImplementedError(
+                        "You can only set a slice on a variable."
+                    )  # pragma: no cover
+
+            elif isinstance(targ, ast.Attribute):
+                if isinstance(targ.value, ast.Name):
+                    # CASE: x.y = value
+                    op = self.environment.map(setattr)
+                    obj = self.process_node(block, targ.value)
+                    idx = Constant(targ.attr)
+                    new_node = Apply([op, obj, idx, anf_node], block.graph)
+                    write(targ.value, new_node)
+                else:
+                    # UNSUPPORTED: f().x = value
+                    raise NotImplementedError(
+                        "You can only set an attribute on a variable."
+                    )  # pragma: no cover
+
+            else:
+                raise NotImplementedError(node.targets)  # pragma: no cover
+
+        write(targ, anf_node)
         return block
 
+    def process_AugAssign(self, block: 'Block',
+                          node: ast.AugAssign) -> 'Block':
+        """Process an augmented assignment `x += y`."""
+        # We just transform it into an Assign node that applies a BinOp,
+        # i.e. the AST for `x = x + y`
+        # This may repeat computations, e.g. for `x[a * a] += 1` it will
+        # create two multiplication nodes, but if we integrate CSE later
+        # that problem will go away on its own.
+        app = ast.BinOp()
+        app.op = node.op
+        app.left = node.target
+        app.right = node.value
+        ass = ast.Assign()
+        ass.targets = [node.target]
+        ass.value = app
+        return self.process_Assign(block, ass)
```

Comment:
I think we should disallow augmented assignment since that too implies inplace modification.